packages/storage/src/reference.ts

```diff
@@ -132,224 +99,281 @@ export class Reference {
+ * "/"s. Firebase Storage List API will filter these unsupported objects.
+ * list() may fail if there are too many unsupported objects in the bucket.
+ * @public
+ *
+ * @param ref - Storage Reference to get list from.
+ * @param options - See ListOptions for details.
+ * @returns A Promise that resolves with the items and prefixes.
+ *      `prefixes` contains references to sub-folders and `items`
+ *      contains references to objects in this folder. `nextPageToken`
+ *      can be used to get the rest of the results.
+ */
+export function list(
+  ref: Reference,
+  options?: ListOptions | null
+): Promise<ListResult> {
+  if (options != null) {
+    listOptionSpec().validator(options);
   }
+  return ref.service.getAuthToken().then(authToken => {
+    const op = options || {};
+    const requestInfo = requests.list(
+      ref.service,
+      ref.location,
+      /*delimiter= */ '/',
+      op.pageToken,
+      op.maxResults
+    );
+    return ref.service.makeRequest(requestInfo, authToken).getPromise();
+  });
+}
 
-  private throwIfRoot_(name: string): void {
-    if (this.location.path === '') {
-      throw errorsExports.invalidRootOperation(name);
-    }
+/**
+ * A promise that resolves with the metadata for this object. If this
+ * object doesn't exist or metadata cannot be retreived, the promise is
+ * rejected.
+ * @public
+ * @param ref - Storage Reference to get metadata from.
+ */
+export function getMetadata(ref: Reference): Promise<Metadata> {
+  ref.throwIfRoot_('getMetadata');
+  return ref.service.getAuthToken().then(authToken => {
+    const requestInfo = requests.getMetadata(
+      ref.service,
+      ref.location,
+      getMappings()
+    );
+    return ref.service.makeRequest(requestInfo, authToken).getPromise();
+  });
+}
+
+/**
+ * Updates the metadata for this object.
+ * @public
+ * @param ref - Storage Reference to update metadata for.
+ * @param metadata - The new metadata for the object.
+ *     Only values that have been explicitly set will be changed. Explicitly
+ *     setting a value to null will remove the metadata.
+ * @returns A promise that resolves
+ *     with the new metadata for this object.
+ *     @see firebaseStorage.Reference.prototype.getMetadata
+ */
+export function updateMetadata(
+  ref: Reference,
+  metadata: Metadata
+): Promise<Metadata> {
+  metadataValidator(metadata);
+  ref.throwIfRoot_('updateMetadata');
+  return ref.service.getAuthToken().then(authToken => {
+    const requestInfo = requests.updateMetadata(
+      ref.service,
+      ref.location,
+      metadata,
+      getMappings()
+    );
+    return ref.service.makeRequest(requestInfo, authToken).getPromise();
+  });
+}
+
+/**
+ * @public
+ * @returns A promise that resolves with the download
+ *     URL for this object.
+ */
+export function getDownloadURL(ref: Reference): Promise<string> {
+  ref.throwIfRoot_('getDownloadURL');
+  return ref.service.getAuthToken().then(authToken => {
```

Comment:
Nit: Use async await to make it more readable. Applies to other methods as well.

Reply:
Done.